controlplane/kubeadm/controllers/scale.go

```diff
@@ -130,6 +143,93 @@ func (r *KubeadmControlPlaneReconciler) scaleDownControlPlane(
+	// so it is considered ok to proceed.
+	if controlPlane.Machines.Len() == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	// If there are deleting machines, wait for the operation to complete.
+	if controlPlane.HasDeletingMachine() {
+		logger.Info("Waiting for machines to be deleted", "Machines", strings.Join(controlPlane.Machines.Filter(machinefilters.HasDeletionTimestamp).Names(), ", "))
+		return ctrl.Result{RequeueAfter: deleteRequeueAfter}, nil
+	}
+
+	// Check machine health conditions; if there are conditions with False or Unknown, then wait.
+	allMachineHealthConditions := []clusterv1.ConditionType{
+		controlplanev1.MachineAPIServerPodHealthyCondition,
+		controlplanev1.MachineControllerManagerPodHealthyCondition,
+		controlplanev1.MachineSchedulerPodHealthyCondition,
+	}
+	if controlPlane.IsEtcdManaged() {
+		allMachineHealthConditions = append(allMachineHealthConditions,
+			controlplanev1.MachineEtcdPodHealthyCondition,
+			controlplanev1.MachineEtcdMemberHealthyCondition,
+		)
+	}
+	machineErrors := []error{}
+	for _, machine := range controlPlane.Machines {
+		for _, condition := range allMachineHealthConditions {
+			if err := preflightCheckCondition(machine, condition); err != nil {
+				machineErrors = append(machineErrors, err)
+			}
+		}
+	}
+	if len(machineErrors) > 0 {
+		aggregatedError := kerrors.NewAggregate(machineErrors)
+		r.recorder.Eventf(controlPlane.KCP, corev1.EventTypeWarning, "ControlPlaneUnhealthy",
+			"Waiting for control plane to pass preflight checks to continue reconciliation: %v", aggregatedError)
+		logger.Info("Waiting for control plane to pass preflight checks", "failures", aggregatedError.Error())
+
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
+
+	// Check KCP conditions ; if there are health problems wait.
+	// NOTE: WE are checking KCP conditions for problems that can't be assigned to a specific machine, e.g.
+	// a control plane node without a corresponding machine
+	allKcpHealthConditions := []clusterv1.ConditionType{
+		controlplanev1.ControlPlaneComponentsHealthyCondition,
+		controlplanev1.EtcdClusterHealthyCondition,
+	}
+	kcpErrors := []error{}
+	for _, condition := range allKcpHealthConditions {
+		if err := preflightCheckCondition(controlPlane.KCP, condition); err != nil {
+			kcpErrors = append(kcpErrors, err)
+		}
+	}
+	if len(kcpErrors) > 0 {
+		aggregatedError := kerrors.NewAggregate(kcpErrors)
+		r.recorder.Eventf(controlPlane.KCP, corev1.EventTypeWarning, "ControlPlaneUnhealthy",
+			"Waiting for control plane to pass preflight checks to continue reconciliation: %v", aggregatedError)
+		logger.Info("Waiting for control plane to pass preflight checks", "failures", aggregatedError.Error())
+
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func preflightCheckCondition(obj conditions.Getter, condition clusterv1.ConditionType) error {
```

Comment:
This is called for KCP conditions too, the error messages should indicate that too?

Reply:
good catch, fixed